src/InputOutput/writers.jl

```diff
@@ -511,6 +531,118 @@ function write!(
     )
 end
 
+"""
+    write!(
+        writer::HDF5Writer,
+        values::DataLayouts.AbstractData,
+        name::AbstractString,
+        topology::Topologies.AbstractTopology,
+    )
+
+Write an object of type 'AbstractData' and name 'name' to the HDF5 file.
+
+The `values` should belong to a `Field` whose `space`'s topology is
+`topology(axes(field))`.
+"""
+function write!(
+    writer::HDF5Writer,
+    group,
+    values::DataLayouts.AbstractData,
+    name::AbstractString,
+    topology::Topologies.AbstractTopology,
+)
+    if topology isa Topologies.Topology2D &&
+       !(writer.context isa ClimaComms.SingletonCommsContext)
+        nelems = Topologies.nelems(topology)
+        (; local_elem_gidx) = topology
+        _write_mpi!(group, values, name; nelems, local_elem_gidx)
+    else
+        _write!(group, values, name)
+    end
+end
+
+function write_plain_array!(group, array::AbstractArray, name::AbstractString)
+    nd = ndims(array)
+    dims = size(array)
+    localidx = ntuple(d -> (:), nd)
+    dataset =
+        create_dataset(group, name, datatype(eltype(array)), dataspace(dims))
+    dataset[localidx...] = array
+    return dataset
+end
+
+"""
+    _write_mpi!(
+        writer::HDF5Writer,
+        data::DataLayouts.AbstractData,
+        name::AbstractString,
+        nelems,
+        local_elem_gidx
+    )
+
+This is an internal method, meant to be used for writing data layouts to the
+HDF5 file.
+
+This method should be used when
+
+```julia
+!(topology isa Topologies.Topology2D && !(writer.context isa ClimaComms.SingletonCommsContext))
+```
+"""
+function _write_mpi!(
+    group,
+    values::DataLayouts.AbstractData,
+    name::AbstractString;
+    nelems,
+    local_elem_gidx,
+)
+    h_dim = DataLayouts.h_dim(DataLayouts.singleton(values))
+    array = parent(values)
+    nd = ndims(array)
+    dims = ntuple(d -> d == h_dim ? nelems : size(array, d), nd)
+    localidx = ntuple(d -> d == h_dim ? local_elem_gidx : (:), nd)
+    dataset = create_dataset(
+        group,
+        "data/$name",
+        datatype(eltype(array)),
+        dataspace(dims);
+        dxpl_mpio = :collective,
+    )
+    dataset[localidx...] = array
+    write_attribute(dataset, "array", array)
+    write_attribute(dataset, "data_layout", string(nameof(typeof(values))))
+    write_attribute(dataset, "data_eltype", string(eltype(values)))
+    return name
+end
+
+"""
+    _write!(
+        writer::HDF5Writer,
+        data::DataLayouts.AbstractData,
+        name::AbstractString,
+    )
+
+This is an internal method, meant to be used for writing data layouts to the
+HDF5 file.
+
+This method should be used when
+
+```julia
+!(topology isa Topologies.Topology2D && !(writer.context isa ClimaComms.SingletonCommsContext))
```

Comment:
I find this hard to read (because of the double/nested negation). I'd suggest

```suggestion
is_distributed = !(writer.context isa ClimaComms.SingletonCommsContext)
!is_distributed || !(topology isa Topologies.Topology2D) 
```